discussions/prog_repo.md

```diff
@@ -0,0 +1,204 @@
+
+### Portability
+
+It's hard to discuss a build system without broaching the topic of portability.
+Luckily, the eBPF portability story has improved fairly recently with the
+introduction of eBPF CO-RE.
+
+Previous issues with portability and the eBPF CO-RE solution are explained
+here:
+
+https://nakryiko.com/posts/bpf-portability-and-co-re/
+
+Because a Kernel Function Marketplace would doubtlessly be a place where
+contributed eBPF programs would be downloaded to run on a variety of kernel
+versions, we propose that the marketplace only contain eBPF CO-RE programs. 
+
+The userspace components of kernel functions pose a separate, complicated
+portability obstacle, which is compounded by the desire of the L3AF project to
+support userspace programs in multiple languages. For the initial version of
+the marketplace, it should be sufficient to build (if a compiled language) and
+unit test for the x86_64 platform. The userspace component should also document
+any installation dependencies it has (e.g., MySQL, Grafana, Python libraries,
+etc.). Contributors to the marketplace would be responsible to provide the
+necessary build scripts and configuration. 
+
+### Build Process
+
+Ideally, we would build on a common image (and kernel version) for all
+contributed eBPF programs. We believe eBPF CO-RE would allow us to do this.
+Contributors would be expected to keep their code building on the most recent
+image used by our build system.
+
+The image used for our build system could be the most recent release of a
+popular Linux server distribution, for example. Here is some information on the
+distros that support eBPF CO-RE be default:
+
+https://github.com/libbpf/libbpf#bpf-co-re-compile-once--run-everywhere
+
+Userspace programs would also ideally build on a common image.
+
+Build artifacts (i.e., eBPF program bytecode and userspace binaries) would be
+stored in a public file storage repository. Users could download the build
+artifacts directory from this repository. In fact, such a repository could also
+be considered the Kernel Function Marketplace, and the source repository could
+be a separate entity.
+
+After building from source, we would then store the build artifacts into an
+package file and sign it. The file would be signed by the L3AF Project.
+
+### Multiple Versions
+
+Regardless of whether the L3AF project builds programs from source or not,
+Kernel Functions should use semantic versioning. The marketplace should then
+host previous versions (in addition to the current version) in some reasonable
+manner.
+
+## Alternative to Hosting and Building Source Code
+
+An alternative to the L3AF project hosting and building contributed source code
+would be for contributors to submit or self-host a signed package containing
+their Kernel Function and any documentation. The L3AF daemon, by default, would
+only download and run Kernel Functions signed by trusted parties.
```

Comment:
Elaborate: how would L3AFD know who the acceptable trusted parties are?  Does it have a set of public keys?  Where would it store them?  How would you update it to add/remove a trusted party? How would you do key rollover?